### PR TITLE
Add OV inference for 3d detection

### PR DESCRIPTION
**Pull request by sovrasov**
### Summary
- PTQ shows 0 accuracy now, additional configuration finetuning is required. One of the root causes is lower than before quantization confidence.
- MAPI wrapper is temporally located in OTX, since the models design is not finalized yet.
- Fix warnings related to 3d detection
- CLI integration tests are enabled

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/common/losses/cross_focal_loss.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def __init__(
 
         self.cls_criterion = cross_sigmoid_focal_loss
 
-    @custom_fwd(cast_inputs=torch.float32)
+    @custom_fwd(cast_inputs=torch.float32, device_type="cuda")
```
Comment (eunwoosh) [inline src/otx/algo/common/losses/cross_focal_loss.py:82]: It's just a question. could you explain why it is necessary?

File: src/otx/algo/object_detection_3d/backbones/monodetr_resnet.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 from torch import nn
 from torchvision.models._utils import IntermediateLayerGetter
 
+from otx.algo.classification.backbones.torchvision import get_model_weights
```
Comment (eunwoosh) [inline src/otx/algo/object_detection_3d/backbones/monodetr_resnet.py]:
```suggestion
from torchvision.models import get_model_weights
```

File: src/otx/algo/object_detection_3d/backbones/monodetr_resnet.py [hunk 3]
```diff
@@ -111,7 +112,7 @@ def __init__(self, name: str, train_backbone: bool, return_interm_layers: bool,
         norm_layer = FrozenBatchNorm2d
         backbone = getattr(torchvision.models, name)(
             replace_stride_with_dilation=[False, False, dilation],
-            pretrained=True,
+            weights=get_model_weights(name).IMAGENET1K_V1,  # the same as pretrained=True
```
Comment (eunwoosh) [inline src/otx/algo/object_detection_3d/backbones/monodetr_resnet.py:115]: If it's same as previous one, then could you explain why it's changed?

File: src/otx/core/model/detection_3d.py [hunk 4]
```diff
@@ -295,3 +216,349 @@ def get_classification_layers(self, prefix: str = "model.") -> dict[str, dict[st
                 num_extra_classes = 6 * sample_model_dim - 5 * incremental_model_dim
                 classification_layers[prefix + key] = {"stride": stride, "num_extra_classes": num_extra_classes}
         return classification_layers
+
+
+class MonoDETRModel(ImageModel):
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py:221]: This class will be moved in to model_api, right?

File: src/otx/core/model/detection_3d.py [hunk 5]
```diff
@@ -295,3 +216,349 @@ def get_classification_layers(self, prefix: str = "model.") -> dict[str, dict[st
...
+
+    __model__ = "mono_3d_det"
+
+    def __init__(self, inference_adapter: InferenceAdapter, configuration: dict[str, Any], preload: bool = False):
+        """Initializes a 3d detection model.
+
+        Args:
+            inference_adapter (InferenceAdapter): inference adapter containing the underlying model.
+            configuration (dict, optional): configuration overrides the model parameters (see parameters() method).
+              Defaults to dict().
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py]:
```suggestion
            configuration (dict): configuration overrides the model parameters (see parameters() method).
```

File: src/otx/core/model/detection_3d.py [hunk 6]
```diff
@@ -295,3 +216,349 @@ def get_classification_layers(self, prefix: str = "model.") -> dict[str, dict[st
...
+    def __init__(
+        self,
+        model_name: str,
+        model_type: str = "mono_3d_det",
+        async_inference: bool = True,
+        max_num_requests: int | None = None,
+        use_throughput_mode: bool = True,
+        model_api_configuration: dict[str, Any] | None = None,
+        metric: MetricCallable = KittiMetric,
+        score_threshold: float = 0.2,
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py:315]: Could you add docstring for `score_threshold`?

File: src/otx/core/model/detection_3d.py [hunk 7]
```diff
@@ -295,3 +216,349 @@ def get_classification_layers(self, prefix: str = "model.") -> dict[str, dict[st
...
+        depth = outputs["depth"]
+        # decode boxes
+        boxes_3d = torch.gather(out_bbox, 1, topk_boxes.repeat(1, 1, 6))  # b, q', 4
+        # heading angle decoding
+        heading = torch.gather(heading, 1, topk_boxes.repeat(1, 1, 24))
+        # depth decoding
+        depth = torch.gather(depth, 1, topk_boxes.repeat(1, 1, 2))
+        # 3d dims decoding
+        size_3d = torch.gather(size_3d, 1, topk_boxes.repeat(1, 1, 3))
+        # 2d boxes of the corners decoding
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py]:
```suggestion
```

File: src/otx/core/model/detection_3d.py [hunk 8]
```diff
@@ -295,3 +216,349 @@ def get_classification_layers(self, prefix: str = "model.") -> dict[str, dict[st
...
+            preload (bool, optional): forces inference adapter to load the model. Defaults to False.
+        """
+        super().__init__(inference_adapter, configuration, preload)
+        self._check_io_number(3, 5)
+
+    def preprocess(self, inputs: dict[str, np.ndarray]) -> tuple[dict[str, Any], ...]:
+        """Preprocesses the input data for the model.
+
+        Args:
+            inputs (dict[str, np.ndarray]): a dict with image, calibration matrix, and image size
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py:241]: How about adding explanation of inputs with more details? It looks a little bit hard to know which values are necessary in `inputs` with current docstring.

File: src/otx/core/model/detection_3d.py [hunk 9]
```diff
@@ -295,3 +216,349 @@ def get_classification_layers(self, prefix: str = "model.") -> dict[str, dict[st
...
+        return {
+            self.image_blob_name: inputs["image"][None],
+            "calib_matrix": inputs["calib"],
+            "img_sizes": inputs["img_size"][None],
+        }, {
+            "original_shape": inputs["image"].shape,
+            "resized_shape": (self.w, self.h, self.c),
+        }
+
+    def _get_inputs(self) -> tuple[list[Any], list[Any]]:
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py:255]: This functions looks almost same as `ImageModel`'s. Is there any reason to override the function?

File: src/otx/core/model/detection_3d.py [hunk 10]
```diff
@@ -295,3 +216,349 @@ def get_classification_layers(self, prefix: str = "model.") -> dict[str, dict[st
...
+            meta (dict[str, Any]): meta information about the input data
+
+        Returns:
+            dict[str, Any]: postprocessed model outputs
+        """
+        result = {}
+        for k in outputs:
+            result[k] = np.copy(outputs[k])
+
+        return result
```
Comment (eunwoosh) [inline src/otx/core/model/detection_3d.py:296]:
```suggestion
        return result
```
Isn't change above possible?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work :) I left some comments. Please take a look.

File: src/otx/algo/common/losses/cross_focal_loss.py (hunk 1 above)
Comment (sovrasov) [reply src/otx/algo/common/losses/cross_focal_loss.py:82]: Please see changes above in the file. "amp.cuda" is now deprecated

File: src/otx/algo/object_detection_3d/backbones/monodetr_resnet.py (hunk 3 above)
Comment (sovrasov) [reply src/otx/algo/object_detection_3d/backbones/monodetr_resnet.py:115]: `pretrained` flag is now deprecated and causes warnings. The weights are the same, you can check the torchvision's implementation

File: src/otx/core/model/detection_3d.py (hunk 4 above)
Comment (sovrasov) [reply src/otx/core/model/detection_3d.py:221]: Right, now we didn't finalize 3d det inference pipeline

File: src/otx/core/model/detection_3d.py (hunk 10 above)
Comment (sovrasov) [reply src/otx/core/model/detection_3d.py:296]: No, it's not. Outputs are owned by infer request.

File: src/otx/core/model/detection_3d.py (hunk 9 above)
Comment (sovrasov) [reply src/otx/core/model/detection_3d.py:255]: The original get_inputs allows only 2D and 4D inputs, which is not the case for this model

File: src/otx/core/model/detection_3d.py (hunk 8 above)
Comment (sovrasov) [reply src/otx/core/model/detection_3d.py:241]: That makes no sense at the moment, since the exact input shape is not yet defined. We can't leave it like it is now

File: src/otx/core/model/detection_3d.py (hunk 6 above)
Comment (sovrasov) [reply src/otx/core/model/detection_3d.py:315]: That is a workaround, so there's no sense to document it. Hyperparameters for 3d detection task should be properly introduced, and then this field can be removed.

**Review by eunwoosh — APPROVED** [review]

**Review by kprokofi — APPROVED** [review]

**Comment by kprokofi** [comment]
LGTM overall, let's refactor post processing after outputs format for 3D OD decision from Geti side.
